### PR TITLE
Added Queue design doc.

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

### CODE REVIEW

File: docs/design/queue.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    Spec QueueSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+    // Current status of Queue
+    // +optional
+    Status QueueStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+type QueueSpec struct {
+    // The weight of queue to share the resources with each other.
+    Weight int32 `json:"weight,omitempty" protobuf:"bytes,1,opt,name=weight"`
```
Comment (hzxuzhonghu) [inline docs/design/queue.md]: What's the meaning of the value? Should  have a clear desc.

File: docs/design/queue.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+type QueueStatus struct {
+    Unknown int32 `json:"running,omitempty" protobuf:"bytes,1,opt,name=running"`
+    Running int32 `json:"running,omitempty" protobuf:"bytes,2,opt,name=running"`
+    Pending int32 `json:"pending,omitempty" protobuf:"bytes,3,opt,name=pending"`
+    Completed int32 `json:"completed,omitempty" protobuf:"bytes,4,opt,name=completed"`
+    Failed int32 `json:"failed,omitempty" protobuf:"bytes,5,opt,name=failed"`
+    Aborted int32 `json:"aborted,omitempty" protobuf:"bytes,6,opt,name=aborted"`
```
Comment (hzxuzhonghu) [inline docs/design/queue.md]: same here? As a normal user, I have no background ctx, but is a common case.

File: docs/design/queue.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
+# Queue
+
+[@k82cn](http://github.com/k82cn); April 17, 2019
+
+## Motivation
+
+`Queue` was introduced in [kube-batch](http://github.com/kubernetes-sigs/kube-batch) long time ago as an internal feature, which makes all jobs are submitted to the same queue, named `default`. As more and more users would like to share resources with each other by queue, this proposal is going to cover primary features of queue achieve that. 
+
+## Function Specificaion
```
Comment (TommyLike) [inline docs/design/queue.md]: s/Specificaion/Specification

File: docs/design/queue.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    Aborted int32 `json:"aborted,omitempty" protobuf:"bytes,6,opt,name=aborted"`
+}
+```
+
+### QueueController
+
+The `QueueController` will manage the lifecycle of queue: 
+
+1. Watching `PodGroup`/`Job` for status
+2. If `Queue` was deleted, also delete all reated `PodGroup`/`Job` in the queue
```
Comment (TommyLike) [inline docs/design/queue.md]: s/reated/related

File: docs/design/queue.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+### Admission Controller
+
+The admission controller will check `PodGroup`/`Job` 's queue when creation:
+
+1. if the queue does not exist, the creation will be rejected
```
Comment (TommyLike) [inline docs/design/queue.md]: Can we inject the default queue?

File: docs/design/queue.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+### Feature Interaction
+
+#### Customized Job/PodGroup
+
+If the `PodGroup` is created by customized controller, the `QueueController` will count those `PodGroup` into `Unknown` status; because `PodGroup` focus on scheduling specification which did not include customized job's status. 
+
+#### cli
+
+Command line is also enhanced for operater engineers. Three sub-commands are introduced as follow:
```
Comment (TommyLike) [inline docs/design/queue.md]: s/operater/operator

File: docs/design/queue.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+`view` command is used to show the detail of a queue, e.g. creation time; the following command will show the detail of queue `myqueue`
+
+```shell
+$ vkctl queue view myqueue
+```
+
+__list__:
+
+`list` command is used to show all available queue to current user
```
Comment (TommyLike) [inline docs/design/queue.md]: s/queue/queues

File: docs/design/queue.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+* Reclaim action: 
+
+  `reclaim` action will go through all queues to reclaim others by `ReclaimableFn`'s return value; the time complexity is `O(n^2)`. In `ReclaimableFn`, both `proportion` and `gang` will take effect: 1. `proportion` makes sure the queue will not be under-used after reclaim, 2. `gang` makes sure the job will not be reclaimed if its `minAvailable` > 1.
+
+* Backfill action:
+
+  When `allocate` action assign resources to each queue, there's a case that ([kube-batch#492](<https://github.com/kubernetes-sigs/kube-batch/issues/492>)) the resources maybe unnecessary idle because of `proportion` plugin: there are one pending job in two queue each, and the deserved resources of each queue can not meet the requirement of their jobs. In such case, `backfill` action will ignore deserved guarantee of queue to fill idle resources as much as possible. This introduces another potential case that the coming smaller job is blocked; this case will be handle be reserved resources of each queue in other project.
```
Comment (TommyLike) [inline docs/design/queue.md]: `` this case will be handle be reserved resources of each queue in other project.`` is a little hard to understand for me:)

File: docs/design/queue.md [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+1. Watching `PodGroup`/`Job` for status
+2. If `Queue` was deleted, also delete all reated `PodGroup`/`Job` in the queue
+
+### Admission Controller
+
+The admission controller will check `PodGroup`/`Job` 's queue when creation:
+
+1. if the queue does not exist, the creation will be rejected
+2. if the queue is releasing, the creation will be also rejected
```
Comment (TommyLike) [inline docs/design/queue.md]: How about the default weight in admission hook

File: docs/design/queue.md (hunk 4 above)
Comment (hzxuzhonghu) [inline docs/design/queue.md]: quick question: who will create the queue, end user?

File: docs/design/queue.md (hunk 2 above)
Comment (TommyLike) [inline docs/design/queue.md]: What are these statuses used for? For instance, what does a running queue stands for?

File: docs/design/queue.md [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+#### Security
+
+TODO: who can view, use, delete/update related queues?
+
+#### Scheduler
+
+* Proportion plugin: 
+
+  Proportion plugin is used to share resource between `Queue`s by weight. The deserved resource of a queue is `(weight/total-weight) * total-resource`. When allocating resources, it will not allocate resource more than its deserved resources.
```
Comment (hzxuzhonghu) [inline docs/design/queue.md]: IIUC, it seems `total-weight` will sum up all the weight of the queues, as the queue increases, the ` (weight/total-weight)` will change. How do scheduler adjust the proportion of resources an old queue occupies ? Let's say the queue resources are all used up, will this trigger a preemt?

File: docs/design/queue.md (hunk 4 above)
Comment (TommyLike) [inline docs/design/queue.md]: Can we achieve this via "OwnerReference" or something?

File: docs/design/queue.md (hunk 10 above)
Comment (hzxuzhonghu) [reply docs/design/queue.md]: Another case, cluster resources may also change

Comment (TommyLike) [reply docs/design/queue.md]: Kube-batch will continue executing those actions every X period.

File: docs/design/queue.md (hunk 5 above)
Comment (k82cn) [reply docs/design/queue.md]: maybe reject it in the first version; we can refer to how prioirtyCass for such kind of default value.

File: docs/design/queue.md (hunk 8 above)
Comment (k82cn) [reply docs/design/queue.md]: should be `by reserved resource` :) we need to support reserved resource of each queue.

File: docs/design/queue.md (hunk 4 above)
Comment (k82cn) [reply docs/design/queue.md]: not finished yet, refer to security/TODO section.

Comment (k82cn) [reply docs/design/queue.md]: no, queue controller should not controll jobs. I'm thinking maybe jobcontroller should delete job when it's queue was gone.

Comment (k82cn) [reply docs/design/queue.md]: We'll let user to define security policy for such cluster level objects.

**Comment by TommyLike** [comment]
@hzxuzhonghu  would you please working on this ? thanks:)
/cc @k82cn

**Comment by hzxuzhonghu** [comment]
Sure, will do

**Pull request merged**